File: Python/Algorithm/N-th_Root.py
# Complexity: O(n*log(num*(10^d))); for n-th root of number num with d decimal accuracy
# Returns: n-th root of a number with 5 decimal accuracy
# Built-in: pow(num, 1.0/n)

def mul(mid, n):
    ans = 1.0
    for _ in range(n): ans *= mid
    return ans
def nth_root(n, num):
    lo, hi = 1.0, max(1, num)
    while hi - lo > 1e-6:
        mid = (hi + lo) / 2
        if mul(mid, n) < num: lo = mid
        else: hi = mid
    return round(lo, 5)

File: CPP/Algorithm/N-th_Root.cpp
// Complexity: O(n*log(num*(10^d))); for n-th root of number num with d decimal accuracy
// Returns: n-th root of a number with 5 decimal accuracy
// Built-in: pow(num, 1.0/n)

auto mul(auto mid, auto n) {
    long double ans = 1;
    for (int i = 0; i < n; ++i) ans *= mid;
    return ans;
}
auto rt(auto n, auto num) {
    long double l = 1, h = num;
    while (h - l > 1e-6) {  // 1e-6 means accuracy till 5 decimal places
        long double mid = (h + l) / 2;
        if (mul(mid, n) < num) l = mid;
        else h = mid;
    }
    return l;
}